internal/common/db.go

```diff
@@ -121,19 +121,30 @@ func GetClusterFromDBWhere(db *gorm.DB, eagerLoading EagerLoadingState, includeD
 
 	db = prepareClusterDB(db, eagerLoading, includeDeleted)
 	err := db.Take(&cluster, where...).Error
-	return &cluster, err
+	if err != nil {
+		return nil, err
+	}
+	cluster.HostCount = int64(len(cluster.Hosts))
```

Comment:
Won't `HostCount` be 0 when eager loading is disabled even if there are hosts associated with the cluster? 
It would be consistent with the overall `Cluster` object state but inconsistent with DB.

Maybe GORM hooks, especially `AfterFind` can be helpful (https://gorm.io/docs/hooks.html)?
Or Association Counting (https://gorm.io/docs/associations.html#Count-Associations)?

Reply:
Thanks! It definitely goes down my alley of options. I was looking at eager loading disabled/enabled yesterday and I stumbled upon a problem with a table lock when updating the cluster with disabled eagerloading and me trying to add a piece of code to fetch the host count.
Using the hooks looks like a nicer way of handling this. So far I've managed to instruct gorm to discard this field from the DB schema, since this is a UI specific domain field. :smile:

Reply:
Yeah, that looks like a typical calculated column; having it in the DB and maintaining it is asking for trouble:)

Reply:
Good news is that the hooks for `AfterFind` works like a charm for `common.Cluster` as advertised. Bad news is that it doesn't cover when retrieving multiple Clusters (`[]*common.Cluster`), such as the case for `ListClusters`. I tried creating a `type Clusters []*Cluster` and use that type to define its `AfterFind` method to overcome this, but the call was not triggered during the tests. And I'm still having problems with the lock for update with `AfterFind` though, and that's not easily fixable without moving to gorm v2.

`Failed to retrieve total host count for cluster 6fde0e8d-36e1-458c-92d9-94650aaafc87: pq: FOR UPDATE is not allowed with aggregate functions `

If we were using gorm 1.20 or newer (gorm v2 as it is [called](https://gorm.io/docs/v2_release_note.html)) I could use the [clauses](https://github.com/go-gorm/gorm/blob/master/clause/locking_test.go#L17) to probably overcome the update use case, but sadly we're stuck with the previous version for now.

I've tried to migrate locally to v2 and I found, as expected, plenty of things that need to be reworked in order to successfully migrate to v2.

Just to bear in mind, currently it will populate its value with len(cluster.Hosts), which is not always matching the totality of the hosts associated to the cluster, for instance [here](https://github.com/openshift/assisted-service/blob/master/internal/common/db.go#L94-L98), where the eager loading is disabled to fetch all the hosts except those that are in disabled state.

The only way I can see this working for all cases is to populate the `HostCount` at the business layer (inventory.go) for each method that returns a `Cluster` outside the transaction section. It's a brittle approach, honestly.

Should we then settle to break consistency in the model and only assign the `HostCount` for the `ListClusters` use case? 

I'm open to suggestions :smile:

Reply:
Shouldn't this field be filled on host adding/deleting instead of in GET calls so it is always accurate? How do you treat disabled hosts? Do they count?

Reply:
That would work: adding a trigger that updates the `host_count` field in the cluster table every time a new host is added.

@filanov WDYT? we add the field into the db schema and update it via hook (AfterCreate) every time a new host is inserted?

Reply:
> How do you treat disabled hosts? Do they count?
Not at the moment, but after following up with @jtomasek, it turns out the UI expects:
* Ready hosts
* Enabled hosts
* All discovered hosts (a.k.a. total hosts)

Only `Total Hosts` is shown by default, and there is a popover that includes the breakdown.

I'm going to go with @slaviered's suggestion and add 3 fields into the cluster schema and add the hooks to populate them during `AfterSave` (to cover for create, update and delete).